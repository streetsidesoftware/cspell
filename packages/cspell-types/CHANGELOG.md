# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [5.5.2](https://github.com/streetsidesoftware/cspell/compare/v5.5.1...v5.5.2) (2021-05-30)


### Bug Fixes

* Ignore hash links like `[abcdefa]` ([#1293](https://github.com/streetsidesoftware/cspell/issues/1293)) ([888e25d](https://github.com/streetsidesoftware/cspell/commit/888e25d48254cc57c9fb4f3c5a65d57dfffacff0))





## [5.5.1](https://github.com/streetsidesoftware/cspell/compare/v5.5.0...v5.5.1) (2021-05-29)

**Note:** Version bump only for package @cspell/cspell-types





# [5.5.0](https://github.com/streetsidesoftware/cspell/compare/v5.4.1...v5.5.0) (2021-05-29)


### Features

* Remove incorrect Ignore Hex Digits Regexp ([#1277](https://github.com/streetsidesoftware/cspell/issues/1277)) ([2621eb0](https://github.com/streetsidesoftware/cspell/commit/2621eb02f487d9e466b4936bde8650c338b320b8)), closes [#1276](https://github.com/streetsidesoftware/cspell/issues/1276) [#1276](https://github.com/streetsidesoftware/cspell/issues/1276)





## [5.4.1](https://github.com/streetsidesoftware/cspell/compare/v5.4.0...v5.4.1) (2021-05-11)

**Note:** Version bump only for package @cspell/cspell-types





# [5.4.0](https://github.com/streetsidesoftware/cspell/compare/v5.3.12...v5.4.0) (2021-05-05)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.12](https://github.com/streetsidesoftware/cspell/compare/v5.3.11...v5.3.12) (2021-04-06)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.11](https://github.com/streetsidesoftware/cspell/compare/v5.3.10...v5.3.11) (2021-04-03)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.10](https://github.com/streetsidesoftware/cspell/compare/v5.3.9...v5.3.10) (2021-04-02)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.8](https://github.com/streetsidesoftware/cspell/compare/v5.3.7...v5.3.8) (2021-03-17)


### Bug Fixes

* update schema generator ([#1074](https://github.com/streetsidesoftware/cspell/issues/1074)) ([528b7d0](https://github.com/streetsidesoftware/cspell/commit/528b7d0b67e1f947832de24e7b66c3d7dfd5fb7f))





## [5.3.7](https://github.com/streetsidesoftware/cspell/compare/v5.3.7-alpha.3...v5.3.7) (2021-03-05)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.7-alpha.3](https://github.com/streetsidesoftware/cspell/compare/v5.3.7-alpha.2...v5.3.7-alpha.3) (2021-03-05)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.5](https://github.com/streetsidesoftware/cspell/compare/v5.3.4...v5.3.5) (2021-03-05)


### Bug Fixes

* fix schema pattern for language IDs. ([#1034](https://github.com/streetsidesoftware/cspell/issues/1034)) ([591dde2](https://github.com/streetsidesoftware/cspell/commit/591dde2532a7698f41c5bf824ead226a7ae78cfe))





## [5.3.4](https://github.com/streetsidesoftware/cspell/compare/v5.3.3...v5.3.4) (2021-03-01)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.3](https://github.com/streetsidesoftware/cspell/compare/v5.3.2...v5.3.3) (2021-02-26)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.2](https://github.com/streetsidesoftware/cspell/compare/v5.3.1...v5.3.2) (2021-02-26)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.1](https://github.com/streetsidesoftware/cspell/compare/v5.3.0...v5.3.1) (2021-02-25)

**Note:** Version bump only for package @cspell/cspell-types





# [5.3.0](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.4...v5.3.0) (2021-02-25)

**Note:** Version bump only for package @cspell/cspell-types





# [5.3.0-alpha.4](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.3...v5.3.0-alpha.4) (2021-02-25)


### Bug Fixes

* be able to report on glob source. ([#1001](https://github.com/streetsidesoftware/cspell/issues/1001)) ([1020d56](https://github.com/streetsidesoftware/cspell/commit/1020d56ba04e8418d162d4bde8e4948546a8f7d2))





# [5.3.0-alpha.3](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.2...v5.3.0-alpha.3) (2021-02-23)


### Bug Fixes

* Improve reporting on files matching glob patterns. ([#994](https://github.com/streetsidesoftware/cspell/issues/994)) ([da991f9](https://github.com/streetsidesoftware/cspell/commit/da991f93a061c5b64ce437332c7107ef2ef89472))





# [5.3.0-alpha.2](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.1...v5.3.0-alpha.2) (2021-02-22)


### Bug Fixes

* Add schema support for `enableFiletypes`. ([#979](https://github.com/streetsidesoftware/cspell/issues/979)) ([d18706e](https://github.com/streetsidesoftware/cspell/commit/d18706e3ffd7a6446396d339b85a753b4ae451e0))
* Add support for custom dictionaries ([#982](https://github.com/streetsidesoftware/cspell/issues/982)) ([196921d](https://github.com/streetsidesoftware/cspell/commit/196921dc635dce7cfdd5b6c300af54beeebde91a))





# [5.3.0-alpha.1](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.0...v5.3.0-alpha.1) (2021-02-19)


### Bug Fixes

* Fix schema generation to use `deprecatedMessage` ([#972](https://github.com/streetsidesoftware/cspell/issues/972)) ([492dca9](https://github.com/streetsidesoftware/cspell/commit/492dca91466773bdf247fdb87f93d64914d5e3e1))





# [5.3.0-alpha.0](https://github.com/streetsidesoftware/cspell/compare/v5.2.4...v5.3.0-alpha.0) (2021-02-18)


### Features

* Be able to specify files to spell check within the config. ([#948](https://github.com/streetsidesoftware/cspell/issues/948)) ([23f7a48](https://github.com/streetsidesoftware/cspell/commit/23f7a488ef500fb1df5cd234c7d3c2ab4ec02961)), closes [#571](https://github.com/streetsidesoftware/cspell/issues/571)
* Part 1 - glob patterns are relative to the config file. ([#921](https://github.com/streetsidesoftware/cspell/issues/921)) ([a250448](https://github.com/streetsidesoftware/cspell/commit/a2504484ec38f15804cc0a203317266f83566b7c))
* Support local configuration files ([#966](https://github.com/streetsidesoftware/cspell/issues/966)) ([0ccc5fe](https://github.com/streetsidesoftware/cspell/commit/0ccc5fe9eb70ca3a4c6e5a3fc0b653465e76983c))





## [5.2.4](https://github.com/streetsidesoftware/cspell/compare/v5.2.3...v5.2.4) (2021-01-28)

**Note:** Version bump only for package @cspell/cspell-types

<!---
cspell: ignore abcdefa
--->
