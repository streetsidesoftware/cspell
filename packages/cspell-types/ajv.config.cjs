'use strict';

const keywords = ['deprecationMessage', 'markdownDescription', 'scope'];

function addKeywords(ajv) {
    for (const keyword of keywords) {
        ajv.addKeyword(keyword);
    }
}

module.exports = addKeywords;
