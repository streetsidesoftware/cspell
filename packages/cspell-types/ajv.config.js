'use strict';

const keywords = ['deprecatedMessage', 'markdownDescription', 'scope'];

function addKeywords(ajv) {
    for (const keyword of keywords) {
        ajv.addKeyword(keyword);
    }
}

module.exports = addKeywords;
