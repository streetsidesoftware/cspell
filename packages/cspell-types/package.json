{
  "name": "@cspell/cspell-types",
  "publishConfig": {
    "access": "public"
  },
  "version": "6.6.0",
  "description": "Types for cspell and cspell-lib",
  "type": "commonjs",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "typesVersions": {
    "*": {
      "*": [
        "dist/*"
      ],
      "dist/index.d.ts": [
        "dist/index.d.ts"
      ],
      "index.d.ts": [
        "dist/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": "./dist/index.js",
    "./Parser": "./dist/Parser.js",
    "./Parser.js": "./dist/Parser.js",
    "./cspell.schema.json": "./cspell.schema.json",
    "./ajv.config": "./ajv.config.js",
    "./ajv.config.js": "./ajv.config.js"
  },
  "scripts": {
    "clean": "rimraf dist coverage .tsbuildinfo",
    "build": "pnpm run compile && pnpm run build-schema",
    "build-dev": "tsc -p tsconfig.dev.json",
    "build-schema": "ts-json-schema-generator --no-top-ref --path src/CSpellSettingsDef.ts --type CSpellSettings --validation-keywords markdownDescription  --validation-keywords scope --validation-keywords deprecated --validation-keywords deprecationMessage -o  ./cspell.schema.json && cp ./cspell.schema.json ../..",
    "clean-build": "pnpm run clean && pnpm run build",
    "compile": "tsc -p .",
    "test-schema": "ajv -s ./cspell.schema.json -d \"cspell.test.{json,yaml}\" -c ./ajv.config.js",
    "test": "jest && pnpm run test-schema",
    "prepare": "echo Skip Building Types"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/streetsidesoftware/cspell.git"
  },
  "files": [
    "dist",
    "ajv.config.js",
    "cspell.schema.json",
    "!**/__mocks__",
    "!**/*.test.*",
    "!**/*.spec.*",
    "!**/*.map"
  ],
  "keywords": [
    "cspell",
    "types"
  ],
  "author": "Jason Dent",
  "license": "MIT",
  "engines": {
    "node": ">=14"
  },
  "devDependencies": {
    "ajv-cli": "^5.0.0",
    "jest": "^28.1.3",
    "rimraf": "^3.0.2",
    "ts-json-schema-generator": "^1.0.0",
    "typescript": "^4.7.4"
  }
}
