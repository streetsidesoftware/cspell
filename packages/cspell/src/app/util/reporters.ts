import assert from 'node:assert';

import {
    type CSpellReporter,
    type CSpellReporterModule,
    type CSpellSettings,
    type FeaturesSupportedByReporter,
    type FileSettings,
    type Issue,
    IssueType,
    type ProgressFileComplete,
    type ReporterConfiguration,
    type ReporterSettings,
    type ReportIssueOptions,
    type RunResult,
    unknownWordsChoices,
} from '@cspell/cspell-types';
import { dynamicImport } from '@cspell/dynamic-import';

import { pkgDir } from '../pkgInfo.js';
import { ApplicationError, toError } from './errors.js';
import { LintFileResult } from './LintFileResult.js';
import { clean } from './util.js';

export type FinalizedReporter = Required<CSpellReporter>;

function filterFeatureIssues(
    features: FeaturesSupportedByReporter | undefined,
    issue: Issue,
    reportOptions?: ReportIssueOptions,
): boolean {
    if (issue.issueType === IssueType.directive) {
        return (features?.issueType && reportOptions?.validateDirectives) || false;
    }
    if (features?.unknownWords) {
        return true;
    }
    if (!reportOptions) {
        // If no options are provided, report the issue.
        return true;
    }
    // The reporter doesn't support the `unknownWords` feature. Handle the logic here.
    let reportIssue = issue.isFlagged;
    reportIssue ||= !reportOptions.unknownWords || reportOptions.unknownWords === unknownWordsChoices.ReportAll;
    reportIssue ||= reportOptions.unknownWords === unknownWordsChoices.ReportSimple && issue.hasSimpleSuggestions;
    reportIssue ||=
        reportOptions.unknownWords === unknownWordsChoices.ReportCommonTypos && issue.hasPreferredSuggestions;

    return reportIssue || false;
}

function handleIssue(reporter: CSpellReporter, issue: Issue, reportOptions?: ReportIssueOptions | undefined): void {
    if (!reporter.issue) return;
    if (!filterFeatureIssues(reporter.features, issue, reportOptions)) {
        // The reporter does not want to handle this issue.
        return;
    }
    if (!reporter.features?.contextGeneration && !issue.context) {
        issue = { ...issue };
        issue.context = issue.line; // Ensure context is always set if the reporter does not support context generation.
    }
    return reporter.issue(issue, reportOptions);
}

/**
 * Loads reporter modules configured in cspell config file
 */
export async function loadReporters(
    reporters: FileSettings['reporters'],
    defaultReporter: CSpellReporter,
    config: ReporterConfiguration,
): Promise<ReadonlyArray<CSpellReporter>> {
    async function loadReporter(reporterSettings: ReporterSettings): Promise<CSpellReporter | undefined> {
        if (reporterSettings === 'default') return defaultReporter;
        if (!Array.isArray(reporterSettings)) {
            reporterSettings = [reporterSettings];
        }
        const [moduleName, settings] = reporterSettings;

        try {
            const { getReporter }: CSpellReporterModule = await dynamicImport(moduleName, [process.cwd(), pkgDir]);
            return getReporter(settings, config);
        } catch (e: unknown) {
            throw new ApplicationError(`Failed to load reporter ${moduleName}: ${toError(e).message}`);
        }
    }

    reporters = !reporters || !reporters.length ? ['default'] : [...reporters];

    const loadedReporters = await Promise.all(reporters.map(loadReporter));
    return loadedReporters.filter((v: CSpellReporter | undefined): v is CSpellReporter => v !== undefined);
}

export function finalizeReporter(reporter: undefined): undefined;
export function finalizeReporter(reporter: CSpellReporter): FinalizedReporter;
export function finalizeReporter(reporter: CSpellReporter | undefined): FinalizedReporter | undefined;
export function finalizeReporter(reporter: CSpellReporter | undefined): FinalizedReporter | undefined {
    if (!reporter) return undefined;

    if (reporterIsFinalized(reporter)) {
        return reporter;
    }

    const final: FinalizedReporter = {
        issue: (...params) => reporter.issue?.(...params),
        info: (...params) => reporter.info?.(...params),
        debug: (...params) => reporter.debug?.(...params),
        progress: (...params) => reporter.progress?.(...params),
        error: (...params) => reporter.error?.(...params),
        result: (...params) => reporter.result?.(...params),
        features: reporter.features,
    };

    return final;
}

function reporterIsFinalized(reporter: CSpellReporter | FinalizedReporter | undefined): reporter is FinalizedReporter {
    return (
        (!!reporter &&
            reporter.features &&
            typeof reporter.issue === 'function' &&
            typeof reporter.info === 'function' &&
            typeof reporter.debug === 'function' &&
            typeof reporter.error === 'function' &&
            typeof reporter.progress === 'function' &&
            typeof reporter.result === 'function') ||
        false
    );
}

type ReportIssueOptionsKeyMap = Required<{ [K in keyof ReportIssueOptions]: K }>;

const reportIssueOptionsKeyMap: ReportIssueOptionsKeyMap = {
    unknownWords: 'unknownWords',
    validateDirectives: 'validateDirectives',
    showContext: 'showContext',
};

function setValue<K extends keyof ReportIssueOptions>(
    options: ReportIssueOptions,
    key: K,
    value: ReportIssueOptions[K] | undefined,
) {
    if (value !== undefined) {
        options[key] = value;
    }
}

export function extractReporterIssueOptions(settings: CSpellSettings | ReportIssueOptions): ReportIssueOptions {
    const src: ReportIssueOptions = settings as ReportIssueOptions;
    const options: ReportIssueOptions = {} as ReportIssueOptions;
    for (const key in reportIssueOptionsKeyMap) {
        const k = key as keyof ReportIssueOptionsKeyMap;
        setValue(options, k, src[k]);
    }
    return options;
}

export function mergeReportIssueOptions(
    a: ReportIssueOptions | CSpellSettings,
    b: ReportIssueOptions | undefined,
): ReportIssueOptions {
    const options: ReportIssueOptions = extractReporterIssueOptions(a);
    if (!b) return options;
    for (const key in reportIssueOptionsKeyMap) {
        const k = key as keyof ReportIssueOptions;
        setValue(options, k, b[k]);
    }
    return options;
}

export class LintReporter {
    #reporters: FinalizedReporter[] = [];
    #config: ReporterConfiguration;
    #finalized: boolean = false;

    constructor(
        readonly defaultReporter: CSpellReporter,
        config: ReporterConfiguration,
    ) {
        this.#config = config;
        if (defaultReporter) {
            this.#reporters.push(finalizeReporter(defaultReporter));
        }
    }

    get config(): ReporterConfiguration {
        return this.#config;
    }

    set config(config: ReporterConfiguration) {
        assert(!this.#finalized, 'Cannot change the configuration of a finalized reporter');
        this.#config = config;
    }

    issue(issue: Issue, reportOptions?: ReportIssueOptions): void {
        for (const reporter of this.#reporters) {
            handleIssue(reporter, issue, reportOptions);
        }
    }

    info(...params: Parameters<FinalizedReporter['info']>): void {
        for (const reporter of this.#reporters) {
            reporter.info(...params);
        }
    }

    debug(...params: Parameters<FinalizedReporter['debug']>): void {
        for (const reporter of this.#reporters) {
            reporter.debug(...params);
        }
    }

    error(...params: Parameters<FinalizedReporter['error']>): void {
        for (const reporter of this.#reporters) {
            reporter.error(...params);
        }
    }

    progress(...params: Parameters<FinalizedReporter['progress']>): void {
        for (const reporter of this.#reporters) {
            reporter.progress(...params);
        }
    }

    async result(result: RunResult): Promise<void> {
        await Promise.all(this.#reporters.map((reporter) => reporter.result?.(result)));
    }

    get features(): FeaturesSupportedByReporter {
        return {
            unknownWords: true,
            issueType: true,
        };
    }

    async loadReportersAndFinalize(reporters: FileSettings['reporters']): Promise<void> {
        assert(!this.#finalized, 'Cannot change the configuration of a finalized reporter');
        const loaded = await loadReporters(reporters, this.defaultReporter, this.config);
        this.#reporters = [...new Set(loaded)].map((reporter) => finalizeReporter(reporter));
    }

    emitProgressBegin(filename: string, fileNum: number, fileCount: number) {
        this.progress({
            type: 'ProgressFileBegin',
            fileNum,
            fileCount,
            filename,
        });
    }

    emitProgressComplete(filename: string, fileNum: number, fileCount: number, result: LintFileResult) {
        const filteredIssues = result.issues.filter((issue) =>
            filterFeatureIssues({}, issue, result.reportIssueOptions),
        );
        const numIssues = filteredIssues.length;

        for (const reporter of this.#reporters) {
            const progress: ProgressFileComplete = clean({
                type: 'ProgressFileComplete',
                fileNum,
                fileCount,
                filename,
                elapsedTimeMs: result.elapsedTimeMs,
                processed: result.processed,
                numErrors: numIssues || result.errors,
                cached: result.cached,
                perf: result.perf,
                issues: reporter.features && result.issues,
                reportIssueOptions: reporter.features && result.reportIssueOptions,
            });
            reporter.progress(progress);
        }

        // Show the spelling errors after emitting the progress.
        result.issues.forEach((issue) => this.issue(issue, result.reportIssueOptions));

        return numIssues;
    }
}
