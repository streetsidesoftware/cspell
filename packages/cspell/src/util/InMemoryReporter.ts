import type { CSpellReporter, Issue, ProgressItem, RunResult } from '@cspell/cspell-types';

export interface InMemoryResult {
    log: string[];
    issues: Issue[];
    runResult: RunResult | undefined;
}

/**
 * Simple reporter for test purposes
 */
export class InMemoryReporter implements CSpellReporter, InMemoryResult {
    log: string[] = [];
    errors: Error[] = [];
    issueCount = 0;
    errorCount = 0;
    debugCount = 0;
    infoCount = 0;
    progressCount = 0;
    issues: Issue[] = [];
    runResult: RunResult | undefined;

    issue = (issue: Issue): void => {
        this.issues.push(issue);
        this.issueCount += 1;
        const { uri, row, col, text } = issue;
        this.log.push(`Issue: ${uri}[${row}, ${col}]: Unknown word: ${text}`);
    };

    error = (message: string, error: Error): void => {
        this.errorCount += 1;
        this.errors.push(error);
        this.log.push(`Error: ${message} ${error.toString()}`);
    };

    info = (message: string): void => {
        this.infoCount += 1;
        this.log.push(`Info: ${message}`);
    };

    debug = (message: string): void => {
        this.debugCount += 1;
        this.log.push(`Debug: ${message}`);
    };

    progress = (p: ProgressItem): void => {
        this.progressCount += 1;
        this.log.push(`Progress: ${p.type} ${p.fileNum} ${p.fileCount} ${p.filename}`);
    };

    result = (r: RunResult): void => {
        this.runResult = r;
    };

    dump = (): InMemoryResult => ({ log: this.log, issues: this.issues, runResult: this.runResult });
}
