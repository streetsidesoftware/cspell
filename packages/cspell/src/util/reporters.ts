import type {
    CSpellReporter,
    CSpellReporterModule,
    FileSettings,
    ReporterSettings,
    RunResult,
} from '@cspell/cspell-types';

import { ApplicationError, toError } from './errors';

type StandardEmitters = Omit<CSpellReporter, 'result'>;

function callAll<P0>(methods: ((p: P0) => void)[]): (p: P0) => void;
function callAll<P0, P1>(methods: ((p0: P0, p1: P1) => void)[]): (p0: P0, p1: P1) => void;
function callAll<P>(methods: ((...p: P[]) => void)[]): (...p: P[]) => void {
    return (...p: P[]) => {
        for (const method of methods) {
            method(...p);
        }
        return;
    };
}

function extractEmitter<K extends keyof StandardEmitters>(
    reporters: ReadonlyArray<StandardEmitters>,
    emitterName: K
): StandardEmitters[K][] {
    // The `bind` is used in case the reporter is a class.
    return reporters.map((r) => r[emitterName].bind(r) as StandardEmitters[K]);
}

function mergeResultEmitters(reporters: ReadonlyArray<CSpellReporter>): CSpellReporter['result'] {
    return async (result: RunResult) => {
        await Promise.all(reporters.map((reporter) => reporter.result(result)));
    };
}

/**
 * Mergers several cspell reporters into a single one
 */
export function mergeReporters(...reporters: ReadonlyArray<CSpellReporter>): CSpellReporter {
    return {
        issue: callAll(extractEmitter(reporters, 'issue')),
        info: callAll(extractEmitter(reporters, 'info')),
        debug: callAll(extractEmitter(reporters, 'debug')),
        progress: callAll(extractEmitter(reporters, 'progress')),
        error: callAll(extractEmitter(reporters, 'error')),
        result: mergeResultEmitters(reporters),
    };
}

function loadReporter(reporterSettings: ReporterSettings): CSpellReporter | undefined {
    if (!Array.isArray(reporterSettings)) {
        reporterSettings = [reporterSettings];
    }
    const [moduleName, settings] = reporterSettings;

    try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const { getReporter }: CSpellReporterModule = require(moduleName);
        return getReporter(settings);
    } catch (e: unknown) {
        throw new ApplicationError(`Failed to load reporter ${moduleName}: ${toError(e).message}`);
    }
}

/**
 * Loads reporter modules configured in cspell config file
 */
export function loadReporters({ reporters = [] }: Pick<FileSettings, 'reporters'>): ReadonlyArray<CSpellReporter> {
    return reporters.map(loadReporter).filter((v: CSpellReporter | undefined): v is CSpellReporter => v !== undefined);
}
