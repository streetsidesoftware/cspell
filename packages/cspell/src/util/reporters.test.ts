import type { CSpellReporter, ReporterSettings } from '@cspell/cspell-types';
import { MessageTypes } from '@cspell/cspell-types';

import { InMemoryReporter } from './InMemoryReporter';
import { loadReporters, mergeReporters } from './reporters';

const defaultReporter: CSpellReporter = {
    issue: jest.fn(),
    info: jest.fn(),
    debug: jest.fn(),
    error: jest.fn(),
    progress: jest.fn(),
    result: jest.fn(),
};

describe('mergeReporters', () => {
    it('processes a single reporter', async () => {
        const reporter = new InMemoryReporter();
        await runReporter(mergeReporters(reporter));

        expect(reporter.dump()).toMatchSnapshot();
    });

    it('processes a multiple reporters', async () => {
        const reporters = [new InMemoryReporter(), new InMemoryReporter(), new InMemoryReporter()];
        await runReporter(mergeReporters(...reporters));

        reporters.forEach((reporter) => {
            expect(reporter.dump()).toMatchSnapshot();
        });
    });

    test('loadReporters', () => {
        const reporters: ReporterSettings[] = [['@cspell/cspell-json-reporter', { outFile: 'out.json' }]];
        const loaded = loadReporters({ reporters }, defaultReporter, {});
        expect(loaded).toEqual([expect.objectContaining({})]);
    });

    test.each`
        reporter                               | expected
        ${['@cspell/cspell-json-reporter']}    | ${'Failed to load reporter @cspell/cspell-json-reporter: cspell-json-reporter settings must be an object'}
        ${['@cspell/cspell-unknown-reporter']} | ${"Failed to load reporter @cspell/cspell-unknown-reporter: Cannot find module '@cspell/cspell-unknown-reporter' from 'src/util/reporters.ts'"}
        ${'@cspell/cspell-unknown-reporter'}   | ${"Failed to load reporter @cspell/cspell-unknown-reporter: Cannot find module '@cspell/cspell-unknown-reporter' from 'src/util/reporters.ts'"}
    `('loadReporters fail $reporter', ({ reporter, expected }) => {
        const reporters: ReporterSettings[] = [reporter];
        const fn = () => loadReporters({ reporters }, defaultReporter, {});
        expect(fn).toThrow(expected);
    });
});

async function runReporter(reporter: CSpellReporter): Promise<void> {
    reporter.debug?.('foo');
    reporter.debug?.('bar');

    reporter.error?.('something went wrong', new Error('oh geez'));

    reporter.info?.('some logs', MessageTypes.Info);
    reporter.info?.('some warnings', MessageTypes.Warning);
    reporter.info?.('some debug logs', MessageTypes.Debug);

    // cSpell:disable
    reporter.issue?.({
        text: 'fulll',
        offset: 13,
        line: { text: 'This text is fulll of errrorrrs.', offset: 0 },
        row: 1,
        col: 14,
        uri: 'text.txt',
        context: { text: 'This text is fulll of errrorrrs.', offset: 0 },
    });
    // cSpell:enable

    reporter.progress?.({
        type: 'ProgressFileComplete',
        fileNum: 1,
        fileCount: 1,
        filename: 'text.txt',
        elapsedTimeMs: 349.058747,
        processed: true,
        numErrors: 2,
    });

    await reporter.result?.({
        files: 1,
        filesWithIssues: new Set(['text.txt']),
        issues: 2,
        errors: 1,
        cachedFiles: 0,
    });
}
