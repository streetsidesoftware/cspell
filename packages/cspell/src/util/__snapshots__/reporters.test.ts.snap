// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeReporters processes a multiple reporters 1`] = `
{
  "issues": [
    {
      "col": 14,
      "context": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "line": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "offset": 13,
      "row": 1,
      "text": "fulll",
      "uri": "text.txt",
    },
  ],
  "log": [
    "Debug: foo",
    "Debug: bar",
    "Error: something went wrong Error: oh geez",
    "Info: some logs",
    "Info: some warnings",
    "Info: some debug logs",
    "Issue: text.txt[1, 14]: Unknown word: fulll",
    "Progress: ProgressFileComplete 1 1 text.txt",
  ],
  "runResult": {
    "cachedFiles": 0,
    "errors": 1,
    "files": 1,
    "filesWithIssues": Set {
      "text.txt",
    },
    "issues": 2,
  },
}
`;

exports[`mergeReporters processes a multiple reporters 2`] = `
{
  "issues": [
    {
      "col": 14,
      "context": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "line": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "offset": 13,
      "row": 1,
      "text": "fulll",
      "uri": "text.txt",
    },
  ],
  "log": [
    "Debug: foo",
    "Debug: bar",
    "Error: something went wrong Error: oh geez",
    "Info: some logs",
    "Info: some warnings",
    "Info: some debug logs",
    "Issue: text.txt[1, 14]: Unknown word: fulll",
    "Progress: ProgressFileComplete 1 1 text.txt",
  ],
  "runResult": {
    "cachedFiles": 0,
    "errors": 1,
    "files": 1,
    "filesWithIssues": Set {
      "text.txt",
    },
    "issues": 2,
  },
}
`;

exports[`mergeReporters processes a multiple reporters 3`] = `
{
  "issues": [
    {
      "col": 14,
      "context": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "line": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "offset": 13,
      "row": 1,
      "text": "fulll",
      "uri": "text.txt",
    },
  ],
  "log": [
    "Debug: foo",
    "Debug: bar",
    "Error: something went wrong Error: oh geez",
    "Info: some logs",
    "Info: some warnings",
    "Info: some debug logs",
    "Issue: text.txt[1, 14]: Unknown word: fulll",
    "Progress: ProgressFileComplete 1 1 text.txt",
  ],
  "runResult": {
    "cachedFiles": 0,
    "errors": 1,
    "files": 1,
    "filesWithIssues": Set {
      "text.txt",
    },
    "issues": 2,
  },
}
`;

exports[`mergeReporters processes a single reporter 1`] = `
{
  "issues": [
    {
      "col": 14,
      "context": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "line": {
        "offset": 0,
        "text": "This text is fulll of errrorrrs.",
      },
      "offset": 13,
      "row": 1,
      "text": "fulll",
      "uri": "text.txt",
    },
  ],
  "log": [
    "Debug: foo",
    "Debug: bar",
    "Error: something went wrong Error: oh geez",
    "Info: some logs",
    "Info: some warnings",
    "Info: some debug logs",
    "Issue: text.txt[1, 14]: Unknown word: fulll",
    "Progress: ProgressFileComplete 1 1 text.txt",
  ],
  "runResult": {
    "cachedFiles": 0,
    "errors": 1,
    "files": 1,
    "filesWithIssues": Set {
      "text.txt",
    },
    "issues": 2,
  },
}
`;
